File: server/src/stats.js
function mean(values) {
  if (!values || values.length === 0) {
    return null;
  }
  const total = values.reduce((sum, value) => sum + value, 0);
  return total / values.length;
}

function variance(values) {
  if (!values || values.length === 0) {
    return null;
  }
  const avg = mean(values);
  const total = values.reduce((sum, value) => {
    const diff = value - avg;
    return sum + diff * diff;
  }, 0);
  return total / values.length;
}

function round(value, digits = 2) {
  if (value === null || value === undefined) {
    return null;
  }
  const factor = Math.pow(10, digits);
  return Math.round(value * factor) / factor;
}

function computeStats(series, fields) {
  const stats = {};
  fields.forEach((field) => {
    const values = series
      .map((entry) => entry[field])
      .filter((value) => typeof value === "number" && !Number.isNaN(value));
    stats[field] = {
      mean: round(mean(values)),
      variance: round(variance(values)),
    };
  });
  return stats;
}

module.exports = {
  mean,
  variance,
  round,
  computeStats,
};

File: server/stats.py
from __future__ import annotations

import math
from typing import Iterable


def mean(values: Iterable[float]) -> float | None:
    values = list(values)
    if not values:
        return None
    return sum(values) / len(values)


def variance(values: Iterable[float]) -> float | None:
    values = list(values)
    if not values:
        return None
    avg = mean(values)
    if avg is None:
        return None
    return sum((value - avg) ** 2 for value in values) / len(values)


def std(values: Iterable[float]) -> float | None:
    var = variance(values)
    if var is None:
        return None
    return math.sqrt(var)


def round_value(value: float | None, digits: int = 2) -> float | None:
    if value is None:
        return None
    factor = 10**digits
    return round(value * factor) / factor


def compute_stats(series: list[dict], fields: list[str]) -> dict[str, dict[str, float | None]]:
    stats: dict[str, dict[str, float | None]] = {}
    for field in fields:
        values = [
            entry.get(field)
            for entry in series
            if isinstance(entry, dict) and isinstance(entry.get(field), (int, float))
        ]
        stats[field] = {
            "mean": round_value(mean(values)),
            "variance": round_value(variance(values)),
            "std": round_value(std(values)),
        }
    return stats
